src/irmin-pack/unix/append_only_file.ml

```diff
@@ -0,0 +1,127 @@
+(*
+ * Copyright (c) 2022-2022 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Import
+include Append_only_file_intf
+
+module Make (Io : Io.S) = struct
+  module Io = Io
+
+  type rw_perm = {
+    buf : Buffer.t;
+    auto_flush_threshold : int;
+    auto_flush_callback : unit -> unit;
+  }
+
+  type t = {
+    io : Io.t;
+    mutable persisted_end_offset : int63;
+    dead_header_size : int63;
+    rw_perm : rw_perm option;
+  }
+
+  let create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_callback =
+    let open Result_syntax in
+    let+ io = Io.create ~path ~overwrite in
+    let persisted_end_offset = Int63.zero in
+    let buf = Buffer.create 0 in
+    {
+      io;
+      persisted_end_offset;
+      dead_header_size = Int63.zero;
+      rw_perm = Some { buf; auto_flush_threshold; auto_flush_callback };
+    }
+
+  let open_rw ~path ~end_offset ~dead_header_size ~auto_flush_threshold
```

Comment:
could use a comment to say that these `open` functions are called after opening the control file, and we read the end_offset from there